File: app/src/main/java/com/aatthhuurr/mobisuas_2411500025/RiwayatActivity.java
package com.aatthhuurr.mobisuas_2411500025;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.ImageButton;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.textfield.TextInputEditText;
import java.util.ArrayList;
import java.util.List;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RiwayatActivity extends AppCompatActivity {

    private ImageButton btnBack;
    private TextInputEditText edtSearch;
    private MaterialButton btnFilter;
    private RecyclerView rvRiwayat;
    private FloatingActionButton fabTambahLaporan;

    private RiwayatAdapter adapter;
    private List<LaporanModel> listAsli = new ArrayList<>();
    private List<LaporanModel> listFilter = new ArrayList<>();

    // 1. Deklarasi variabel ID User dipindah ke sini (sebagai variabel global class)
    private String idUserLogedIn = "1";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_riwayat);

        btnBack = findViewById(R.id.btnBack);
        edtSearch = findViewById(R.id.edtSearch);
        btnFilter = findViewById(R.id.btnFilter);
        rvRiwayat = findViewById(R.id.rvRiwayat);
        fabTambahLaporan = findViewById(R.id.fabTambahLaporan);

        // 2. Logika menangkap Intent SEKARANG SUDAH MASUK ke dalam onCreate() agar tidak error
        if (getIntent().hasExtra("ID_USER")) {
            idUserLogedIn = getIntent().getStringExtra("ID_USER");
        }

        rvRiwayat.setLayoutManager(new LinearLayoutManager(this));

        // Ambil data dari API berdasarkan ID User yang aktif
        loadLaporanDariServer();

        fabTambahLaporan.setOnClickListener(v -> {
            startActivity(new Intent(RiwayatActivity.this, PelaporanActivity.class));
        });

        // Logika Pencarian Real-Time judul laporan
        edtSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                cariJudulLaporan(s.toString());
            }
            @Override
            public void afterTextChanged(Editable s) {}
        });

        btnBack.setOnClickListener(v -> finish());
    }

    private void loadLaporanDariServer() {
        ApiClient.getClient().getRiwayat(idUserLogedIn).enqueue(new Callback<List<LaporanModel>>() {
            @Override
            public void onResponse(Call<List<LaporanModel>> call, Response<List<LaporanModel>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    listAsli.clear();
                    listAsli.addAll(response.body());
                    listFilter.clear();
                    listFilter.addAll(listAsli);

                    adapter = new RiwayatAdapter(RiwayatActivity.this, listFilter);
                    rvRiwayat.setAdapter(adapter);
                }
            }
            @Override
            public void onFailure(Call<List<LaporanModel>> call, Throwable t) {
                Toast.makeText(RiwayatActivity.this, "Gagal memuat riwayat", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void cariJudulLaporan(String query) {
        listFilter.clear();
        for (LaporanModel item : listAsli) {
            if (item.getJudul().toLowerCase().contains(query.toLowerCase())) {
                listFilter.add(item);
            }
        }
        if (adapter != null) adapter.notifyDataSetChanged();
    }
}

File: app/src/main/java/com/aatthhuurr/mobisuas_2411500025/RiwayatActivity.kt
package com.aatthhuurr.mobisuas_2411500025

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.ImageButton
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.aatthhuurr.mobisuas_2411500025.api.RetrofitClient
import com.aatthhuurr.mobisuas_2411500025.helper.SharedPrefManager
import com.aatthhuurr.mobisuas_2411500025.model.LaporanResponse
import com.google.android.material.button.MaterialButton
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.textfield.TextInputEditText
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class RiwayatActivity : AppCompatActivity() {

    private lateinit var swipeRefresh: SwipeRefreshLayout
    private lateinit var rvRiwayat: RecyclerView
    private lateinit var prefManager: SharedPrefManager
    private var idUserTerlogin: Int = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_riwayat)

        val btnBack = findViewById<ImageButton>(R.id.btnBack)
        val edtSearch = findViewById<TextInputEditText>(R.id.edtSearch)
        val fabTambahLaporan = findViewById<FloatingActionButton>(R.id.fabTambahLaporan)

        rvRiwayat = findViewById(R.id.rvRiwayat)
        swipeRefresh = findViewById(R.id.swipeRefresh)

        rvRiwayat.layoutManager = LinearLayoutManager(this)

        prefManager = SharedPrefManager(this)
        idUserTerlogin = prefManager.getIdUser()

        // Pemicu tarik manual (Swipe to Refresh) tetap aktif
        swipeRefresh.setOnRefreshListener {
            checkAndLoadData()
        }

        fabTambahLaporan.setOnClickListener {
            startActivity(Intent(this, PelaporanActivity::class.java))
        }

        btnBack.setOnClickListener {
            // 1. Bersihkan seluruh session login di Shared Preference
            prefManager.logout()

            Toast.makeText(this, "Berhasil keluar dari akun", Toast.LENGTH_SHORT).show()

            // 2. Arahkan kembali ke halaman Login dengan menghancurkan seluruh tumpukan activity belakang (Clear Task)
            val intent = Intent(this, LoginActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            startActivity(intent)

            // 3. Tutup halaman riwayat secara permanen
            finish()
        }
    }

    override fun onResume() {
        super.onResume()
        checkAndLoadData() // Otomatis refresh data tanpa perlu ditarik manual
    }
    // =========================================================================

    private fun checkAndLoadData() {
        if (idUserTerlogin != -1) {
            muatDataRiwayatLaporan(idUserTerlogin, rvRiwayat)
        } else {
            swipeRefresh.isRefreshing = false
            Toast.makeText(this, "Sesi bermasalah, silakan login ulang", Toast.LENGTH_SHORT).show()
        }
    }

    private fun muatDataRiwayatLaporan(idUser: Int, recyclerView: RecyclerView) {
        RetrofitClient.instance.getRiwayat(idUser).enqueue(object : Callback<LaporanResponse> {
            override fun onResponse(call: Call<LaporanResponse>, response: Response<LaporanResponse>) {
                swipeRefresh.isRefreshing = false

                if (response.isSuccessful && response.body() != null) {
                    val res = response.body()!!

                    if (res.status) {
                        val listData = res.data
                        val adapter = RiwayatAdapter(listData) { laporan ->
                            val intent = Intent(this@RiwayatActivity, DetailRiwayatActivity::class.java)
                            intent.putExtra("ID_LAPORAN", laporan.id_laporan)
                            startActivity(intent)
                        }
                        recyclerView.adapter = adapter
                        Log.d("API_SUCCESS", "Jumlah data laporan: ${listData.size}")
                    } else {
                        Toast.makeText(this@RiwayatActivity, res.message, Toast.LENGTH_SHORT).show()
                    }
                } else {
                    Toast.makeText(this@RiwayatActivity, "Gagal memproses data server", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<LaporanResponse>, t: Throwable) {
                swipeRefresh.isRefreshing = false
                Log.e("API_ERROR", "Error: ${t.message}")
                Toast.makeText(this@RiwayatActivity, "Gagal memuat riwayat: ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }
}